### ISSUE TITLE
Update background-subtraction intermediate products for WFSS

### ISSUE THREAD

**Issue by stscijgbot-jp** [spacetelescope/jwst#8608]
_Issue [JP-3673](https://jira.stsci.edu/browse/JP-3673) was created on JIRA by [Alicia Canipe](https://jira.stsci.edu/secure/ViewProfile.jspa?name=acanipe):_

<https://jira.stsci.edu/browse/JP-898> was completed as part of the work for JP-897.

WFSS experts have reported that their community would find an intermediate product (somewhere between the rate.fits files and the cal.fits files) to be more useful for their own science and data processing than the current pipeline outputs. We could consider saving the products from JP-898 in MAST by default, or discuss another option (e.g., saving rate.fits files plus assign_wcs and flat_field). WFSS experts should confirm/weigh in.

Additional request (Dec 04 2024): One of the steps of the WFSS background subtraction is to mask the traces to only scale the background on sky regions without traces. This mask is an important parameter, as an improper masking would bias the scaling. This mask (the output of 'background_sub.mask_from_source_cat(...)') is not saved/stored anywhere presently, and it would be useful to have so users can check if the masking is appropriate or not. It could be saved as a new extension of the "_bsub.fits" file (the background subtracted image) or a new fits file altogether. This might be the same output than the "contam_model" of the wfss_contam step but this is not clear presently (since different functions are used), but even if it were, ideally the user shouldn't have to run an extra/later step to get the mask of the background_sub step.

---------

February 2025 update:

Discussed at <https://outerspace.stsci.edu/display/JWSTPWG/2025-Feb-13+WFSS+Pipeline+Discussion;> consensus to move ahead on this ticket.  I've modified the original ticket to reflect the latest agreed-upon behavior:

Plan is to write out intermediate files from the background subtraction step.  These will be in units of DN/s.  The files should include:
 * The name of the subtracted background file.  This is retrieved from CRDS and should already be in the header, but I don't see it currently.  Per <https://jira.stsci.edu/projects/CRDS/issues/CRDS-920> this will presumably need a new R_BKGD (or whatever other phrase is more appropriate) keyword.
 * An extension giving the derived source mask
 * Per <https://jira.stsci.edu/browse/JP-3808,> a header keyword giving the scaling factor applied to the reference background during the subtraction process.

We should also ensure that it's possible to pass this intermediate file back into the pipeline starting at the next step (flat_field) so that any edits to the file get picked up and processed.  (I.e., allowing users to customize the mask and background scaling in some offline manner, and then resume pipeline processing at that point).

Ultimately it looks like we'll want to enable creation of these products by default, but it may be preferable to have them OFF by default in the code and enable them via parameter ref files after we're happy with the testing.

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-3673?focusedCommentId=707046#comment-707046):_

Pinging [Rachel Plesha](https://jira.stsci.edu/secure/ViewProfile.jspa?name=rplesha) to comment for NIRISS.

Also, following #6783 I presume this intermediate product should include background subtraction as well.

**Comment by stscijgbot-jp** [comment]
_Comment by [Rachel Plesha](https://jira.stsci.edu/secure/ViewProfile.jspa?name=rplesha) on [JIRA](https://jira.stsci.edu/browse/JP-3673?focusedCommentId=713400#comment-713400):_

The NIRISS WFSS team discussed this ticket. We agree that there are some intermediate products that would be useful to save:
1) assign_wcs + flat_field products. This is useful for display purposes and getting access to sky coords
2) single file/extension with list of all source bounding box information when looking at the grism image. This should include source ID, starting and ending x/y locations (for both grism and direct images), and the source type.
3) Residual measurements for contamination modeling

For the assign_wcs & flat field products, we discussed if the background should be included or not. For now, we do not think that this should be included, but we will want to revisit this after more analysis is done on the background calibration (ongoing now). We believe we can easily update this in the future by setting the --save_combined_background parameter to true in a parameter reference file.

For the bounding box information, this is something that is available in cal and x1d datamodel metadata (output from extract_2d), but is a little tedious to get. JP-3169 has a bit more information about how this information could be input into the x1d headers, but another option we discussed was adding some columns into the source catalog as well. The end goal here is to be able to easily go between the direct image+source catalog to the 2D and 1D spectra for any given source.

For the contamination modeling residual measurements, we think this is part of a larger discussion about the contamination modeling, and it appears there is an option of "--save_simulated_image" that can be set if we decide we need this in the future. I am mentioning it here more for record keeping than an action item.

**Comment by stscijgbot-jp** [comment]
_Comment by [Gael Noirot](https://jira.stsci.edu/secure/ViewProfile.jspa?name=gnoirot) on [JIRA](https://jira.stsci.edu/browse/JP-3673?focusedCommentId=759328#comment-759328):_

Just adding an additional request here (description above updated to reflect this request):
One of the steps of the WFSS background subtraction is to mask the traces to only scale the background on sky regions without traces. This mask is an important parameter, as an improper masking would bias the scaling. This mask (the output of 'background_sub.mask_from_source_cat(...)') is not saved/stored anywhere presently, and it would be useful to have so users can check if the masking is appropriate or not. It could be saved as a new extension of the “_bsub.fits” file (the background subtracted image) or a new fits file altogether. This might be the same output than the "contam_model" of the wfss_contam step but this is not clear presently (since different functions are used), but even if it were, ideally the user shouldn't have to run an extra/later step to get the mask of the background_sub step.

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-3673?focusedCommentId=929924#comment-929924):_

Looking back at this ticket and trying to further nail down some details.

As I see it, the reason why these intermediate products are desirable is because 1) Rate files show the entire detector, but have significant flatfield/background illumination artifacts and do not have any embedded WCS information.  2) Cal files fix both of those issues with the rate files, but have been sliced up into lots of pieces of the field for individual objects by extract2d.  Given how integral the source-finding is to WFSS processing, it makes sense to provide a product with all relevant corrections just before all of the cutting-up is done.  The goal here is to allow users to do more in the way of custom source extraction (whether in our pipeline or outside of it), without the need to start all the way back at the rate file stage with running assign_wcs.

After that general concept we get into the weeds, with regard to units of the products (MJy/sr vs flatfielded DN/s vs flatfielded e-/sec), whether or not the background subtraction and/or flatfielding steps have been applied, and how source information is provided.

I'd suggest the following:
 * We produce by default (and save in MAST) a single additional file after the bkg-sub and flatfield steps.  No reason users should need to rerun flatfielding just to change the cutouts.  It should be possible to condense what's needed into a single file to minimize the additional data products being provided.
 * A header keyword, e.g. BKGD_SCL) records the scaling applied to the background reference file (see <https://jira.stsci.edu/browse/JP-3808)>
 * An extension of this file saves the background that was subtracted from the image.
 * Another extension of this file saves the mask that was used when determining the background scaling.
 * From a user perspective, it would then be simple to look at the mask to determine how good it was.  If not good enough, the saved as-subtracted background image combined with the BKGD_SCL keyword would suffice to reconstruct the original data to be modified accordingly.
 * Ideally, it should be possible to pass this intermediate product into the spec2 pipeline starting at the extract_2d stage to continue running the rest of the pipeline after tweaking the intermediate product and potentially the source catalog as necessary.

Still unclear in my mind:
 * This data product seems closer to the _esec.fits file than the _bsub.fits file since it would be written after the flatfield step.  Is it a problem that the flatfield has been applied if trying to redo the background subtraction?  If so does the flatfield need to be available to back out the original data?  If so, is it sufficient for users to retrieve the flatfield from CRDS or should it also be saved as an extension in this product?
 * What units should this product be, flatfielded DN/s or flatfielded e-/s?  DN/s seems cleaner from the perspective of being able to pass the modified input back into the extract2d pipeline again, although the very name of the current optional product is 'esec'
 * What is the aim of providing source ID, start/stop locations, source types, etc in a single file?  Is it just for information purposes, or to be something that can be modified to affect the processing?  I.e., I presume some of this information isn't determined until after the extract_2d and srctype steps?  Is source information entirely attached to the 'esec' file or does the extract_2d step still rely on the i2d, cat, and segm files in the extract_2d step to compute that?  Is JP-3751 to record this information in the x1d files all that's necessary?
 * What kind of residual measurements are desirable?  Anything that isn't already contained in the (image - background*scale), (background*scale), and (mask) extensions?

Thoughts [Rachel Plesha](https://jira.stsci.edu/secure/ViewProfile.jspa?name=rplesha) [Gael Noirot](https://jira.stsci.edu/secure/ViewProfile.jspa?name=gnoirot) [Russell Ryan](https://jira.stsci.edu/secure/ViewProfile.jspa?name=rryan) ?

**Comment by stscijgbot-jp** [comment]
_Comment by [Rachel Plesha](https://jira.stsci.edu/secure/ViewProfile.jspa?name=rplesha) on [JIRA](https://jira.stsci.edu/browse/JP-3673?focusedCommentId=930056#comment-930056):_

Thanks for summarizing that, David! I agree with everything you've summarized.

I think the only question I can speak to now is about the source ID start/stop locations. When we talked about this, we envisioned this to be used for informational purposes only, and I think that JP-3751 would solve this. The goal was to easily be able to tell what source you are looking at when looking at a spectrum, which is a bit roundabout right now.

I think this is separate but similar to a request that [Russell Ryan](https://jira.stsci.edu/secure/ViewProfile.jspa?name=rryan) has voiced which is to have a table of all positions and wavelengths for all sources which is to help with the contamination modeling. I'll let him speak to that as I can't find my notes about this currently to confirm.

**Comment by stscijgbot-jp** [comment]
_Comment by [Gael Noirot](https://jira.stsci.edu/secure/ViewProfile.jspa?name=gnoirot) on [JIRA](https://jira.stsci.edu/browse/JP-3673?focusedCommentId=939331#comment-939331):_

Hi [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw), thanks a lot for summarizing all of those points.

I am wondering if saving the "background*scaling" image is really necessary. The background is a file available in the CRDS cache which can be retrieved by any user who wants to do advanced processing or checks, similarly to the flat field image. What really needs to be saved and is not available somewhere else is the mask and the scaling factor. From those two any user can check if the masking is appropriate, and if not they can, if they want, recreate the non-bcksub image using the scaling factor plus the CRDS background.
So I believe, only saving the mask image and the scaling factor should be enough, and can be done before the flat-fielding step, together with the _bsub.fits file for instance.

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-3673?focusedCommentId=945235#comment-945235):_

Discussed at <https://outerspace.stsci.edu/display/JWSTPWG/2025-Feb-13+WFSS+Pipeline+Discussion;> consensus to move ahead on this ticket.  I've modified the original ticket to reflect the latest agreed-upon behavior:

Plan is to write out intermediate files from the background subtraction step.  These will be in units of DN/s.  The files should include:
 * The name of the subtracted background file.  This is retrieved from CRDS and should already be in the header, but I don't see it currently.  Per <https://jira.stsci.edu/projects/CRDS/issues/CRDS-920> this will presumably need a new R_BKGD (or whatever other phrase is more appropriate) keyword.
 * An extension giving the derived source mask
 * Per <https://jira.stsci.edu/browse/JP-3808,> a header keyword giving the scaling factor applied to the reference background during the subtraction process.

We should also ensure that it's possible to pass this intermediate file back into the pipeline starting at the next step (flat_field) so that any edits to the file get picked up and processed.  (I.e., allowing users to customize the mask and background scaling in some offline manner, and then resume pipeline processing at that point).

Ultimately it looks like we'll want to enable creation of these products by default, but it may be preferable to have them OFF by default in the code and enable them via parameter ref files after we're happy with the testing.

**Comment by stscijgbot-jp** [comment]
_Comment by [Maria Pena-Guerrero](https://jira.stsci.edu/secure/ViewProfile.jspa?name=pena) on [JIRA](https://jira.stsci.edu/browse/JP-3673?focusedCommentId=957015#comment-957015):_

[David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) we need further clarification, please. Which of these options do you mean?
 # Save the background mask as an optional file to be modified and inputed back into the background step only
 # Have the background saved into the final product as an extension so that it can be used at a later stage in pipeline for WFSS processing.

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-3673?focusedCommentId=957201#comment-957201):_

[Maria Pena-Guerrero](https://jira.stsci.edu/secure/ViewProfile.jspa?name=pena) Neither.  What I meant is that the background-subtraction step should save the background-subtracted data to a file (along with the usual SCI/ERR/DQ etc extensions).  That file should also contain an extension with the as-applied boolean mask indicating where sources were masked out during the background subtraction process.  There should also be a header keyword containing the scaling factor that was applied to the reference background during the subtraction process.

The intent here is that users could download this file, make alterations to the background subtraction if they wished, and then resume spec2 processing with the modified file.  We don't need to include the reference background in the data file since it already exists in CRDS, just the mask and the scale factor.

**Comment by stscijgbot-jp** [comment]
_Comment by [Rachel Plesha](https://jira.stsci.edu/secure/ViewProfile.jspa?name=rplesha) on [JIRA](https://jira.stsci.edu/browse/JP-3673?focusedCommentId=958574#comment-958574):_

We were chatting about this ticket today at a NIRISS WFSS meeting and noticed that the flat-field component of this has seemingly been dropped from the plans. Originally the goal of this product was to introduce a way for users to have a quick-look grism image in the full frame with assign_wcs & flatfield applied. We then rightfully threw in that users often modify the background as well, so we should include information that would make this easy to do. However, this throws in some complications because the flat field step is applied after background. 

Several folks on the NIRISS WFSS team still feel that having a flat fielded product for the quick look is valuable. This could certainly use some discussion (perhaps in a couple of weeks when folks are back from vacation?), but we are proposing that we have the flat fielded, background subtracted data in this product in addition to the background mask (in another extension) as this might be a good middle ground for this product to meet both goals of being a quick-look product, but also allowing for easy access to modify the background mask.

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-3673?focusedCommentId=959538#comment-959538):_

[Rachel Plesha](https://jira.stsci.edu/secure/ViewProfile.jspa?name=rplesha) It seems pretty awkward to have the background subtraction step write out flatfielded results, or to have the flatfield step write out internal products from the bg subtraction step.  Looking at some of the discussion on the original tickets though suggests that we might be able to resolve this by moving the flatfield step before the background subtraction step in the pipeline for WFSS since (per [Kevin Volk](https://jira.stsci.edu/secure/ViewProfile.jspa?name=kvolk)) the two should commute.  It seems like there's now a reason to do so; flatfield shouldn't be a step that we expect anyone to modify and rerun, whereas bg subtraction is.

This would of course mean that the background reference files would need to be flatfielded, but that should be an easy update.  Thoughts?

**Comment by stscijgbot-jp** [comment]
_Comment by [Rachel Plesha](https://jira.stsci.edu/secure/ViewProfile.jspa?name=rplesha) on [JIRA](https://jira.stsci.edu/browse/JP-3673?focusedCommentId=959617#comment-959617):_

Would this just be for WFSS data that we switch the order, or would we switch the order for all modes?

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-3673?focusedCommentId=959627#comment-959627):_

Only WFSS; for some other modes the order is more important since background subtraction uses observations rather than reference files.

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-3673?focusedCommentId=960402#comment-960402):_

Split off the discussion of flatfield and background ordering to <https://jira.stsci.edu/browse/JP-4002> so that this ticket should be able to move ahead.

**Comment by stscijgbot-jp** [comment]
_Comment by [Melanie Clarke](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mclarke) on [JIRA](https://jira.stsci.edu/browse/JP-3673?focusedCommentId=960892#comment-960892):_

Setting to on hold until JP-4002 is resolved. When that's done, we'll still need some clarity on what exactly is needed for the expected processing flow and saved output data format.

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-3673?focusedCommentId=969819#comment-969819):_

Per discussion with the WFSS team 7/9/25:
 * Order of flatfield and background subtraction will not be changed.  Background subtraction happens first, flatfield second.
 * The intermediate product should be written out by the background subtraction step.  I.e., it will not be flatfielded.  This is ok because the flatfield variations will not be very apparent anyway since they will not be noticeable in the background sky (which has been subtracted to zero) and just manifest as some slightly fainter traces and slightly different statistics.
 * The intermediate product should be '_bsub' and should include the mask used by the background subtraction step in an extension, along with the scaling factor used in a header keyword.

In the possibility that MIRI may at some point need wavelength-dependent flatfields, this approach should work fine.  MIRI would simply skip the regular flatfield step and apply something fancier at the x1d level.  Details TBD in future, if this proves necessary.

Separately, there is interest in allowing the background subtraction step to accept an optional input mask to be used instead of relying upon an internally-generated mask.  This will be filed as a separate ticket to be linked below.

Science decision is to make these intermediate bsub files be produced by default by DMS and provided to users in MAST.  Edit: This should probably be set to False in the code, and enabled by parameter reference file for WFSS modes to avoid complicating the logic for other instrument modes.

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-3673?focusedCommentId=969820#comment-969820):_

See <https://jira.stsci.edu/browse/JP-4056> for discussion of whether or not the background subtraction step should be able to accept an input mask.  This is undecided at the present time, but discussion can happen on that ticket and does not need to hold up implementation of the present ticket on creating intermediate products.

**Comment by stscijgbot-jp** [comment]
_Comment by [Maria Pena-Guerrero](https://jira.stsci.edu/secure/ViewProfile.jspa?name=pena) on [JIRA](https://jira.stsci.edu/browse/JP-3673?focusedCommentId=972676#comment-972676):_

[David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) I have only saved the background mask in an extension for WFSS data, but in re-reading the comments it is unclear to me if you are asking for the mask to be accessible as well for the other modes. In order to save it as an extension, we added a field for the mask in the image datamodel and so, saving the intermediary results with the regular step.save_results=True will work. No need to add a new parameter. For the future, saving the background mask as an extension for other modes should be as easy as adding a line to assign the new datamodel parameter to the mask, e.g. 
```java
model.mask = bkg_mask ```
where bkg_mask is expected to be a 2D array.

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-3673?focusedCommentId=973260#comment-973260):_

[Maria Pena-Guerrero](https://jira.stsci.edu/secure/ViewProfile.jspa?name=pena) Just writing out the mask for WFSS modes is fine; in general other modes don't use a mask anyway.  Good to hear that save_results=True will work, thanks.

**Comment by stscijgbot-jp** [comment]
_Comment by [Hien Tran](https://jira.stsci.edu/secure/ViewProfile.jspa?name=htran) on [JIRA](https://jira.stsci.edu/browse/JP-3673?focusedCommentId=974321#comment-974321):_

[Maria Pena-Guerrero](https://jira.stsci.edu/secure/ViewProfile.jspa?name=pena) can you provide some examples of the names of files for the intermediate background-subtracted products?

**Comment by stscijgbot-jp** [comment]
_Comment by [Maria Pena-Guerrero](https://jira.stsci.edu/secure/ViewProfile.jspa?name=pena) on [JIRA](https://jira.stsci.edu/browse/JP-3673?focusedCommentId=974330#comment-974330):_

[Hien Tran](https://jira.stsci.edu/secure/ViewProfile.jspa?name=htran) an example intermediate file name is:
```java
jw04368001001_05101_00008_nis_bsub.fits```
You get this file by running:
```java
strun calwebb_spec2 jw04368-o001_20250619t133612_spec2_00001_asn.json --steps.bkg_subtract.save_results=True```

**Comment by stscijgbot-jp** [comment]
_Comment by [Melanie Clarke](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mclarke) on [JIRA](https://jira.stsci.edu/browse/JP-3673?focusedCommentId=974339#comment-974339):_

A couple of notes about these files:
 * The output filenames will have suffix "bsub", but otherwise follow the same naming patterns as rate or cal files from this mode.
 * As far as I know, "bsub" products have never been saved during automated processing for any mode or instrument in the past, so this will be a new file type from MAST's perspective.
 * Saving the file by default will be triggered by delivering a new parameter reference file for this mode that sets save_results=True for the bkg_subtract step.  This will need to be coordinated between the instrument team and DMS, to make sure that DMS can support the new file type before saving it is turned on.

**Comment by stscijgbot-jp** [comment]
_Comment by [Rachel Plesha](https://jira.stsci.edu/secure/ViewProfile.jspa?name=rplesha) on [JIRA](https://jira.stsci.edu/browse/JP-3673?focusedCommentId=974349#comment-974349):_

-As a quick note, we've discussed a slightly modified suffix possibly to be more immediately descriptive of the file for users. When would we need to get that nailed down?-

 

Based on the discussion Katie linked, it makes the most sense to move forward with just the "bsub" suffix unless we find a strong reason to change this.

**Comment by stscijgbot-jp** [comment]
_Comment by [Katie Kaleida](https://jira.stsci.edu/secure/ViewProfile.jspa?name=ckaleida) on [JIRA](https://jira.stsci.edu/browse/JP-3673?focusedCommentId=974392#comment-974392):_

We met on 2025/08/06 to flesh out the impacts to other DMS components from these changes.  Those meeting minutes can be found here: <https://innerspace.stsci.edu/display/jwstdms/2025-08-06+Impacts+of+JP-3673>

**Comment by stscijgbot-jp** [comment]
_Comment by [Melanie Clarke](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mclarke) on [JIRA](https://jira.stsci.edu/browse/JP-3673?focusedCommentId=974783#comment-974783):_

Background mask added to data product, in a MASK FITS extension, in <https://github.com/spacetelescope/jwst/pull/9698.>

I'll leave this unresolved for now to track the parameter reference file delivery, and assign it back to David to track.

**Comment by stscijgbot-jp** [comment]
_Comment by [Rachel Plesha](https://jira.stsci.edu/secure/ViewProfile.jspa?name=rplesha) on [JIRA](https://jira.stsci.edu/browse/JP-3673?focusedCommentId=976311#comment-976311):_

[Maria Pena-Guerrero](https://jira.stsci.edu/secure/ViewProfile.jspa?name=pena) in testing this, the background mask extension looks great! However, I noticed that if the background subtraction step exits out because it can't find enough pixels to subtract, then the format of the file/the number of extensions changes because the mask is not saved out. We discussed this today, and we think it would be best if the format of the file stays the same, i.e. the mask extension is present even if the background step stops processing. 

 

There was some discussion about if the data in the array should be all 0s or the actual mask itself. Let's start by saving the failed mask if that is possible, and if we find that users start submitting help desk questions being confused about having a mask that wasn't actually applied, we can consider switching it to all zeros. (It will be at least 95% all zeros anyway by definition in the step).

 

It might also make sense to take care of [JP-4062](https://jira.stsci.edu/browse/JP-4062?src=confmacro) at the same time as this fix.

**Comment by stscijgbot-jp** [comment]
_Comment by [Rachel Plesha](https://jira.stsci.edu/secure/ViewProfile.jspa?name=rplesha) on [JIRA](https://jira.stsci.edu/browse/JP-3673?focusedCommentId=976518#comment-976518):_

[Katie Kaleida](https://jira.stsci.edu/secure/ViewProfile.jspa?name=ckaleida) I've created the CRDS tickets with the top level ticket as CRDS-1008, and individually (rmap, file delivery):

NIRISS: CRDS-1009 & CRDS-1012

NIRCam: CRDS-1010 & CRDS-1013

MIRI: CRDS-1011 & CRDS-1014

 

We'll wait for word from MAST before we deliver the files officially. It would be good to know by the end of next week if possible because we have to deliver the files before September 5th.

**Comment by stscijgbot-jp** [comment]
_Comment by [Melanie Clarke](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mclarke) on [JIRA](https://jira.stsci.edu/browse/JP-3673?focusedCommentId=977423#comment-977423):_

Assigning back to [Maria Pena-Guerrero](https://jira.stsci.edu/secure/ViewProfile.jspa?name=pena) to address what happens when the step exits early.  We can assign back to David when that's done.

**Comment by stscijgbot-jp** [comment]
_Comment by [Maria Pena-Guerrero](https://jira.stsci.edu/secure/ViewProfile.jspa?name=pena) on [JIRA](https://jira.stsci.edu/browse/JP-3673?focusedCommentId=977474#comment-977474):_

[Rachel Plesha](https://jira.stsci.edu/secure/ViewProfile.jspa?name=rplesha) what about the scaling factor, do you want that in the header even if the step is skipped or save it as maybe 0 or N/A to mark that the background mask was not applied?

**Comment by stscijgbot-jp** [comment]
_Comment by [Rachel Plesha](https://jira.stsci.edu/secure/ViewProfile.jspa?name=rplesha) on [JIRA](https://jira.stsci.edu/browse/JP-3673?focusedCommentId=977478#comment-977478):_

[Maria Pena-Guerrero](https://jira.stsci.edu/secure/ViewProfile.jspa?name=pena) yes that would be great! We think N/A might be best if that works. That is filed in [JP-4062](https://jira.stsci.edu/browse/JP-4062?src=confmacro)

**Comment by stscijgbot-jp** [comment]
_Comment by [Maria Pena-Guerrero](https://jira.stsci.edu/secure/ViewProfile.jspa?name=pena) on [JIRA](https://jira.stsci.edu/browse/JP-3673?focusedCommentId=977652#comment-977652):_

[Rachel Plesha](https://jira.stsci.edu/secure/ViewProfile.jspa?name=rplesha) and [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw), I looked into it and in order to add a value of N/A for the 
BKGD_SCL keyword, we would have to change the expected type of the datamodel to string and that comes with its own downfalls. As an alternative, we propose that the keyword value is set to 1.0 and the completion status for the step is set to FAILED instead of SKIPPED. The mask would be saved regardless. What do you think?

**Comment by stscijgbot-jp** [comment]
_Comment by [Rachel Plesha](https://jira.stsci.edu/secure/ViewProfile.jspa?name=rplesha) on [JIRA](https://jira.stsci.edu/browse/JP-3673?focusedCommentId=977671#comment-977671):_

That sounds reasonable to me. [Jo Taylor](https://jira.stsci.edu/secure/ViewProfile.jspa?name=jotaylor) & [Alicia Canipe](https://jira.stsci.edu/secure/ViewProfile.jspa?name=acanipe) – any objections to this plan?

**Comment by stscijgbot-jp** [comment]
_Comment by [Jo Taylor](https://jira.stsci.edu/secure/ViewProfile.jspa?name=jotaylor) on [JIRA](https://jira.stsci.edu/browse/JP-3673?focusedCommentId=977715#comment-977715):_

This seems reasonable!

**Comment by stscijgbot-jp** [comment]
_Comment by [Gael Noirot](https://jira.stsci.edu/secure/ViewProfile.jspa?name=gnoirot) on [JIRA](https://jira.stsci.edu/browse/JP-3673?focusedCommentId=978054#comment-978054):_

I would suggest setting the value to 0 since the step would be skipped an no background would be removed, which is equivalent to applying a factor of 0 to the background subtraction.

**Comment by stscijgbot-jp** [comment]
_Comment by [Maria Pena-Guerrero](https://jira.stsci.edu/secure/ViewProfile.jspa?name=pena) on [JIRA](https://jira.stsci.edu/browse/JP-3673?focusedCommentId=978477#comment-978477):_

The new behavior for WFSS is that if the step exists early, the mask will be saved in the corresponding extension, the step status will be set to S_BKDSUB=FAILED, and the BKGD_SCL keyword will be set to 0.0

**Comment by stscijgbot-jp** [comment]
_Comment by [Melanie Clarke](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mclarke) on [JIRA](https://jira.stsci.edu/browse/JP-3673?focusedCommentId=978488#comment-978488):_

PR merged to address early exit behavior:

#9772

Assigning back to David to track parameter reference file delivery, when ready.

**Comment by stscijgbot-jp** [comment]
_Comment by [Rachel Plesha](https://jira.stsci.edu/secure/ViewProfile.jspa?name=rplesha) on [JIRA](https://jira.stsci.edu/browse/JP-3673?focusedCommentId=978502#comment-978502):_

Thanks [Maria Pena-Guerrero](https://jira.stsci.edu/secure/ViewProfile.jspa?name=pena) – I'll test this out next week!

**Comment by stscijgbot-jp** [comment]
_Comment by [Rachel Plesha](https://jira.stsci.edu/secure/ViewProfile.jspa?name=rplesha) on [JIRA](https://jira.stsci.edu/browse/JP-3673?focusedCommentId=978872#comment-978872):_

I looked over the products and things look as expected now – thanks!

When the background subtraction isn't performed, the scale factor is 0, the mask is still saved, and the step is set to FAILED.

I attached a couple of examples showing the masks and scale factors

(Fail): !jw01090001001_03101_00002_nis_bsub_bkgmask.png|width=600!

(Success): !jw03362002001_07101_00001_nis_bsub_bkgmask.png|width=600!

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-3673?focusedCommentId=981154#comment-981154):_

Pars-backgroundstep files for all 3 WFSS modes have been delivered to CRDS-TEST, so setting this to RTT and assigning to the WFSS group to close out once the build is finished.

**Comment by stscijgbot-jp** [comment]
_Comment by [Katie Kaleida](https://jira.stsci.edu/secure/ViewProfile.jspa?name=ckaleida) on [JIRA](https://jira.stsci.edu/browse/JP-3673?focusedCommentId=982799#comment-982799):_

Is PID 04368 the best test data for use to run though on D-string as part of our DMS 12.1 integration (specifically 04368001001)?  Any others you would recommend instead of or in addition to PID 04368?  [Rachel Plesha](https://jira.stsci.edu/secure/ViewProfile.jspa?name=rplesha) et al?

**Comment by stscijgbot-jp** [comment]
_Comment by [Rachel Plesha](https://jira.stsci.edu/secure/ViewProfile.jspa?name=rplesha) on [JIRA](https://jira.stsci.edu/browse/JP-3673?focusedCommentId=983208#comment-983208):_

[Katie Kaleida](https://jira.stsci.edu/secure/ViewProfile.jspa?name=ckaleida)

From the NIRISS perspective, these cases would be good to test, please reach out to [Jo Taylor](https://jira.stsci.edu/secure/ViewProfile.jspa?name=jotaylor) if you need these to be narrowed down more from the full observations over the next couple of days as I'll be out!
 * PID 1089 obs 001 (typical observation)
 * PID 1090 obs 001 (in particular: jw01090001001_03101_00002 & jw01090001001_11101_00001 failed the background subtraction step)
 * PID 1536 obs 008 (WFSS-128[R/C] subarrays)

MIRI does not yet have on-sky data yet. This will be next month, so they don't have any to test right now.

I haven't heard back from NIRCam, but maybe [Alicia Canipe](https://jira.stsci.edu/secure/ViewProfile.jspa?name=acanipe) can comment?

**Comment by stscijgbot-jp** [comment]
_Comment by [Bryan Hilbert](https://jira.stsci.edu/secure/ViewProfile.jspa?name=hilbert) on [JIRA](https://jira.stsci.edu/browse/JP-3673?focusedCommentId=983244#comment-983244):_

If you need NIRCam examples, here are a few that seem like they would make good tests:


 * PID 1076 Obs 117 (crowded field)
 * PID 2078 Obs 016 (sparser field)
 * PID 1187 Obs 021 (sparser field with GRISMC)

**Comment by stscijirahub** [comment]
_This issue is tracked on JIRA as [JP-3673](https://jira.stsci.edu/browse/JP-3673)._
